internal/provisioner/plain/controllers/bundleinstance_controller.go

```diff
@@ -262,12 +283,84 @@ func (r *BundleInstanceReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 		Type:    rukpakv1alpha1.TypeInstalled,
 		Status:  metav1.ConditionTrue,
 		Reason:  rukpakv1alpha1.ReasonInstallationSucceeded,
-		Message: fmt.Sprintf("instantiated bundle %s successfully", bi.Spec.BundleName),
+		Message: fmt.Sprintf("instantiated bundle %s successfully", bundle.GetName()),
 	})
-	bi.Status.InstalledBundleName = bi.Spec.BundleName
+	bi.Status.InstalledBundleName = bundle.GetName()
+
+	if _, err := r.reconcileOldBundles(ctx, oldBundles, bundle); err != nil {
+		meta.SetStatusCondition(&bi.Status.Conditions, metav1.Condition{
+			Type:    rukpakv1alpha1.TypeProgressing,
+			Status:  metav1.ConditionFalse,
+			Reason:  rukpakv1alpha1.ReasonCleanupFailed,
+			Message: err.Error(),
+		})
+		return ctrl.Result{}, err
+	}
+	meta.RemoveStatusCondition(&bi.Status.Conditions, rukpakv1alpha1.TypeProgressing)
+
 	return ctrl.Result{}, nil
 }
 
+// reconcileDesiredBundle is responsible for checking whether the desired
+// Bundle resource that's specified in the BundleInstance parameter's
+// spec.Template configuration is present on cluster, and if not, creates
+// a new Bundle resource matching that desired specification.
+func (r *BundleInstanceReconciler) reconcileDesiredBundle(ctx context.Context, bi *rukpakv1alpha1.BundleInstance) (*rukpakv1alpha1.Bundle, []*rukpakv1alpha1.Bundle, error) {
+	// get the set of Bundle resources that already exist on cluster, and sort
+	// by metadata.CreationTimestamp in the case there's multiple Bundles
+	// that match the label selector.
+	existingBundles, err := util.GetBundlesForBundleInstanceSelector(ctx, r.Client, bi)
+	if err != nil {
+		return nil, nil, err
+	}
+	sort.Sort(util.BundlesByCreationTimestamp(existingBundles))
+
+	// check whether the BI controller has already reached the maximum
+	// generated Bundle limit to avoid hotlooping scenarios.
+	if len(existingBundles) > maxGeneratedBundleLimit {
+		return nil, nil, ErrMaxGeneratedLimit
+	}
```

Comment:
How does deployment handle this with replica sets? Can we just start deleting the oldest bundles we have if we reach the limit? That way, we can always try to create the desired bundle as specified by the BI spec.

Reply:
What happens if we have a bug in the comparison logic such that deleting the Bundle with the oldest creation timestamp just continuously results in new Bundles being generated? I had added this condition check mainly to prevent hotlooping behavior while I was still introducing tests, so I only intended for this to be a short term bandaid. I think we can either remove this logic in this PR, or implement a more robust check that mirrors the deployment implementation to prevent this behavior as a follow-up. Thoughts?

Reply:
I'm good either way. It seems like once we work any kinks out, we shouldn't hotloop because the next reconciliation we'll see that the latest bundle matches the desired template (unless of course that template changes again)